qiskit_experiments/framework/base_experiment.py

```diff
@@ -167,29 +144,144 @@ def _initialize_experiment_data(
 
         return experiment_data._copy_metadata()
 
-    def run_analysis(self, experiment_data, **options) -> ExperimentData:
+    def pre_transpile_action(self, backend: Backend):
```

Comment:
Why is this method needed? I don't see why this is necessary and can't just be done as part of the `circuits` (or `_circuits`) method that experiments already have to implement.

Reply:
I'd like to move towards a seamless and easy schedule management in experiments. Currently, in experiments like Rabi, FineAmp, and Drag we have **the** `schedule` which is attached to a particular gate. However, the experiment circuits often feature other gates (e.g. a y-rotation to map a z-rotation error to a qubit population, etc.). This means that to get a self-consistent calibration framework we need to
* Either add the different schedules manually through options which is what we are currently doing. This is unscalable, complicated, and messy.
* Find an automatic way to attach the schedules to the circuits. Fortunately for us Terra 0.19.0 will allow us to do this with the instruction schedule map in an automatic way through a transpiller pass (thanks to Naoki).

This may justify having a `pre_transpile_action`. I have the following question: Could this transpiller pass be done in the current call to the transpiler with an option `inst_map=inst_map_from_cals`? If so I'm not sure we need `pre_transpile_action` if not then we need a `pre_transpile_action`.

Reply:
Yes, this can be done from everywhere you can access to `self`, i.e. `self.set_transpile_options(inst_map=inst_map_from_cals)`. However, this should be called after you get backend, and you run the transpile. Alternative options would be write something like

```python
class BaseCalibration:
    def circuits(self, backend):
        self.set_transpile_options(inst_map=inst_map_from_cals)
        ...

class MyCalExp(BaseCalibration):
    def circuits(self, backend):
        circus = super().circuits(backend)
        ....
```
this is equivalent to add `pre_transpile_action`. So this is mainly readability issue.

Reply:
i.e. separation of experiment program and dynamic configuration.

Reply:
I'd like to keep the `circuits` method clean of any calibration related variables. The following would be nice:
```python
class BaseCalibrationExperiment(BaseExperiment, ABC):  # as per PR 251

    ...
    
    def pre_transpile_action():
        """Set the instruction schedule map for transpiling."""
        self.set_transpile_options(inst_map = self._cals.default_inst_map)

```
alternatively, without the `pre_transpile_action` hook we could probably do:
```python
class BaseCalibrationExperiment(BaseExperiment, ABC):  # as per PR 251

    def __init__(self, ..., calibrations=cals):
        ...
        self.set_transpile_options(inst_map=cals.default_inst_map)
```
would this second option work? I think it should be okay as `default_inst_map` is a property and is not deep copied around so if I run my calibration experiment twice the update of the cals (and therefore the `default_inst_map`) should also happen. For a regular experiment that does not do any schedule management would the following workflow be sufficient?
```python
exp = MyCoolExperiment(qubits, ...)
exp.set_transpile_options(inst_map=my_backend.defaults().instruction_schedule_map)
exp.run(my_backend)
```